From e7edb589945c005039da76d8040b06128ea43b7d Mon Sep 17 00:00:00 2001
From: Murala S Swamy <ssmurala@habana.ai>
Date: Sun, 30 Jun 2024 08:29:42 +0300
Subject: [PATCH] [FSW-420] fs1: add public key support

with this patch, added public key hash for
pre-production, and 4 key hashes for production.
function support to get key hash and verify key
hash is added.

Change-Id: Ib04c1df59e5cbfa26d6e5653ae13e14a073a7798
---

diff --git a/arch/fs1/arc/dev_security.c b/arch/fs1/arc/dev_security.c
index 58b73d6..31d6b3c 100644
--- a/arch/fs1/arc/dev_security.c
+++ b/arch/fs1/arc/dev_security.c
@@ -11,6 +11,64 @@
 #include <fs1_reg_map.h>
 #include <fs1_efuse.h>
 #include <status.h>
+#include <dev_security.h>
+
+/*
+ * This key hash is calculated on the Public Key Properties -
+ *   -----------------------------------------------
+ * | R ^ 2 | modulus | exponent | inverse | num bits |
+ *   -----------------------------------------------
+ * Since the key props are directly given to the ROM, there is
+ * no need to explicitly include the Public key, which is
+ * usually the modulus.
+ */
+const uint8_t rom_key_hashes[NUM_KEYS_IN_ROM][SHA384_SUM_LEN] = {
+	/* Habana Public Key */
+	{
+		0x95, 0xa3, 0x2a, 0x03, 0xd4, 0xbd, 0xb1, 0xe0,
+		0x2b, 0x54, 0xc0, 0x9c, 0x61, 0x10, 0x80, 0x2c,
+		0x08, 0xc4, 0x5d, 0xdf, 0x4a, 0xf8, 0x03, 0x00,
+		0x2a, 0xf5, 0xac, 0x20, 0x6f, 0x2c, 0x8f, 0x27,
+		0x69, 0x2e, 0x63, 0x7d, 0x02, 0x33, 0x54, 0x29,
+		0x3e, 0x57, 0x46, 0xd9, 0xf4, 0xa9, 0x7e, 0x88
+	},
+	/* Production Key 1 */
+	{
+		0x60, 0x44, 0x16, 0xd8, 0x1b, 0xea, 0x2e, 0x66,
+		0x46, 0xf3, 0x01, 0x78, 0x2f, 0x3c, 0xf6, 0xfc,
+		0x88, 0x3d, 0xbb, 0x5d, 0x94, 0xe7, 0xd9, 0x94,
+		0x79, 0xea, 0x6e, 0xdc, 0x1d, 0xd8, 0x89, 0xe1,
+		0xa4, 0xeb, 0x28, 0x4a, 0x1b, 0x60, 0x87, 0xed,
+		0x58, 0x4a, 0xe1, 0xa0, 0xe9, 0x13, 0x6f, 0xb6
+	},
+	/* Production Key 2 */
+	{
+		0x0c, 0xbe, 0x41, 0xaa, 0x56, 0xd2, 0xe3, 0x29,
+		0x0f, 0x45, 0x2f, 0x48, 0x64, 0x1f, 0x61, 0x00,
+		0xa3, 0x94, 0x6c, 0x2b, 0xc0, 0xa5, 0xd0, 0x9a,
+		0x0d, 0x3d, 0xb3, 0x74, 0x85, 0x1c, 0xbc, 0x9b,
+		0xee, 0xc1, 0x3c, 0xbc, 0x7e, 0xae, 0xf0, 0x7c,
+		0x4a, 0x7e, 0x82, 0x52, 0x69, 0xbc, 0xdc, 0x5f
+	},
+	/* Production Key 3 */
+	{
+		0xd3, 0x21, 0x9d, 0xe5, 0xcf, 0x7e, 0x4c, 0x20,
+		0x98, 0x4e, 0x84, 0x4d, 0x00, 0x5f, 0xf8, 0x71,
+		0x98, 0xf6, 0x11, 0x8f, 0xd4, 0xfd, 0xf6, 0x29,
+		0x6a, 0xd3, 0xfd, 0x94, 0xcf, 0x21, 0xf5, 0x60,
+		0xe8, 0x78, 0x91, 0x4e, 0xc0, 0x51, 0xc6, 0x82,
+		0xe1, 0xae, 0xfe, 0xcf, 0xbf, 0x96, 0xcf, 0xaf
+	},
+	/* Production Key 4 */
+	{
+		0x98, 0xd0, 0x0c, 0xa3, 0xa4, 0xc3, 0xe8, 0x55,
+		0x8d, 0xcf, 0xd2, 0x48, 0x0d, 0x8a, 0x52, 0xd6,
+		0xf5, 0x4c, 0x45, 0x58, 0x6d, 0x20, 0xb9, 0xe2,
+		0x4f, 0xe7, 0x6d, 0x0b, 0xdc, 0x76, 0xf6, 0xad,
+		0x2f, 0x79, 0xca, 0x03, 0x92, 0xc9, 0x2a, 0x5f,
+		0x36, 0x19, 0xee, 0x61, 0x3a, 0x10, 0x69, 0xa9
+	}
+};
 
 void security_load_revoke_bits(void)
 {
diff --git a/arch/fs1/arc/dev_security.h b/arch/fs1/arc/dev_security.h
index ecf5155..4d84965 100644
--- a/arch/fs1/arc/dev_security.h
+++ b/arch/fs1/arc/dev_security.h
@@ -10,4 +10,6 @@
 
 void security_load_revoke_bits(void);
 
+extern const uint8_t rom_key_hashes[NUM_KEYS_IN_ROM][SHA384_SUM_LEN];
+
 #endif /* _DEV_SECURITY_H_ */
diff --git a/common/secure_boot.c b/common/secure_boot.c
index baf2930..f0407d3 100644
--- a/common/secure_boot.c
+++ b/common/secure_boot.c
@@ -7,6 +7,8 @@
 #include <sha512.h>
 #include <rsa.h>
 #include <string.h>
+#include <secure_boot.h>
+#include <security.h>
 
 int secure_boot_verify_image_hash384(uint8_t *exp_hash, unsigned char *img, uint32_t img_len)
 {
@@ -16,3 +18,16 @@
 
 	return memcmp(actual_hash, exp_hash, SHA384_SUM_LEN);
 }
+
+int secure_boot_verify_key_hash(int key_index, uint8_t *key, uint32_t key_len)
+{
+	int ret;
+	uint8_t expected_hash[SHA384_SUM_LEN];
+
+	if (security_rom_get_key_hash_by_idx((uint32_t *)expected_hash, key_index))
+		return SECUREBOOT_VERIFY_KEY_FAILED;
+
+	ret = secure_boot_verify_image_hash384(expected_hash, key, key_len);
+
+	return ret;
+}
diff --git a/common/secure_boot.h b/common/secure_boot.h
index d3263cd..1505674 100644
--- a/common/secure_boot.h
+++ b/common/secure_boot.h
@@ -6,6 +6,12 @@
 #ifndef _SECURE_BOOT_H_
 #define _SECURE_BOOT_H_
 
+#define SECUREBOOT_VERIFY_KEY_REVOKED	-3
+#define SECUREBOOT_VERIFY_KEY_FAILED	-2
+#define SECUREBOOT_VERIFY_VER_FAILED	-1
+#define SECUREBOOT_VERIFY_FAILED	0
+#define SECUREBOOT_VERIFY_SUCCESS	1
+
 int secure_boot_verify_image_hash384(uint8_t *exp_hash, unsigned char *img, uint32_t img_len);
 
 #endif /* _SECURE_BOOT_H_ */
diff --git a/common/security.c b/common/security.c
index c6ef7f0..08b4de9 100644
--- a/common/security.c
+++ b/common/security.c
@@ -7,6 +7,7 @@
 
 #include <boot_rom.h>
 #include <dev_security.h>
+#include <string.h>
 
 int security_check_key_revoke_status(uint32_t key_index)
 {
@@ -47,3 +48,12 @@
 	security_load_sec_efuse_dword();
 	security_load_revoke_bits();
 }
+
+int security_rom_get_key_hash_by_idx(uint32_t *hash, uint8_t key_index)
+{
+	uint32_t *ptr = (uint32_t *)rom_key_hashes[key_index];
+
+	memcpy(hash, ptr, SHA384_SUM_LEN);
+
+	return 0;
+}
diff --git a/common/security.h b/common/security.h
index 7c79fe4..58b9354 100644
--- a/common/security.h
+++ b/common/security.h
@@ -8,5 +8,6 @@
 
 void boot_rom_security_init(void);
 int security_check_key_revoke_status(uint32_t key_index);
+int security_rom_get_key_hash_by_idx(uint32_t *hash, uint8_t key_index);
 
 #endif /* _SECURITY_H_ */
